File: Sample/Engine/Gameplay/Components/ChildActorComponent.py
# -*- encoding: utf-8 -*-

from __future__ import annotations
import builtins
from dataclasses import dataclass
from typing import Any, List, Optional, Tuple, TYPE_CHECKING

from .Component import Component

if TYPE_CHECKING:
    from Engine.Gameplay.Actors import Actor


@Meta(Vector2fVars=["relativePosition"])
@dataclass
class ChildActorComponent(Component):
    r"""\brief Spawn one child actor relative to the owning actor."""

    className: str = ""  #: Actor class path or generated blueprint class name
    relativePosition: Tuple[float, float] = (0.0, 0.0)  #: Spawn position relative to the parent actor

    def __post_init__(self) -> None:
        self._childActor: Optional[Actor] = None

    def init(self, owner: Actor) -> List[Any]:
        r"""
        \brief Spawn and attach the configured child actor.

        - \param owner Actor that owns this component.
        - \return The spawned child actor, or an empty list.
        """
        className = self.className.strip() if isinstance(self.className, str) else ""
        if not className:
            return []

        actorMap = owner.getMap()
        if actorMap is None:
            return []

        childActor = self._childActor
        if childActor is not None and not childActor.isDestroyed():
            return []

        from Source import Data

        childActor = Data.genActorFromClassName(className, self._makeChildTag(owner))
        if childActor is None:
            return []

        owner.addChild(childActor)
        childActor.setRelativePosition(self._normaliseRelativePosition())
        layer = actorMap.getActorLayer(owner)
        if layer is None:
            layer = "default"
        actorMap.spawnActor(childActor, layer, emitCreateEvent=False)
        self._childActor = childActor
        return [childActor]

    def _normaliseRelativePosition(self) -> Tuple[float, float]:
        value = self.relativePosition
        if isinstance(value, str):
            try:
                evaluator = getattr(builtins, "Eval", eval)
                value = evaluator(value)
            except Exception:
                value = (0.0, 0.0)
        if hasattr(value, "x") and hasattr(value, "y"):
            return (float(value.x), float(value.y))
        if isinstance(value, (list, tuple)) and len(value) >= 2:
            try:
                return (float(value[0]), float(value[1]))
            except (TypeError, ValueError):
                pass
        return (0.0, 0.0)

    @staticmethod
    def _makeChildTag(owner: Actor) -> str:
        parentTag = owner.getMapTag()
        if parentTag is None:
            parentTag = ""
        if not isinstance(parentTag, str):
            parentTag = str(parentTag)
        return f"{parentTag}_child"

File: Sample/Scripts/Source/Components/ChildActorComponent.lua
local Engine = require("Engine")

local Component = Engine.Component

local ChildActorComponent = {}

ChildActorComponent.className = ""
ChildActorComponent.relativePosition = sf.Vector2f.new(0.0, 0.0)

function ChildActorComponent:init(values)
    values = values or {}
    self.className = values.className == nil and ChildActorComponent.className or values.className
    local relativePosition = values.relativePosition == nil and ChildActorComponent.relativePosition
        or values.relativePosition
    self.relativePosition = copy(relativePosition)
    ---@type Engine.Actor | nil
    self._childActor = nil
end

function ChildActorComponent:onAttach(owner)
    local className = self.className:match("^%s*(.-)%s*$")
    if not bool(className) then
        return {}
    end
    ---@cast className string

    local actorMap = owner:getMap()
    if actorMap == nil then
        return {}
    end

    if self._childActor ~= nil and not self._childActor:isDestroyed() then
        return {}
    end

    local Data = require("Source.Data")

    local childActor = Data.genActorFromClassName(className, ChildActorComponent._makeChildTag(owner))
    if childActor == nil then
        return {}
    end

    owner:addChild(childActor)
    childActor:setRelativePosition(self.relativePosition)
    local layer = actorMap:getActorLayer(owner)
    if layer == nil then
        layer = "default"
    end
    actorMap:spawnActor(childActor, layer, false)
    self._childActor = childActor
    return { childActor }
end

---@param owner Engine.Actor
---@return string
function ChildActorComponent._makeChildTag(owner)
    local parentTag = owner:getMapTag()
    if parentTag == nil then
        parentTag = ""
    end
    return tostring(parentTag) .. "_child"
end

return class(ChildActorComponent, Component)
